### PR TITLE
[TransferEngine] Bandwidth-aware NIC selection + DMA-BUF and IPv6 binding fixes

### PR DESCRIPTION
**Pull request by Nishikant-K-P**
## Description

Two-commit series. Replaces the legacy `getPciDistance` GPU↔NIC selection
heuristic with NCCL-style PCI ancestor-chain walking and bandwidth-aware
ranking, plus 23 hardware-free unit tests for the new logic. Each commit
is self-contained and bisectable.

> **Note:** This PR was originally a four-commit series. Per reviewer
> request to split, the IPv6 RPC binding fix moved to #2014 and the
> DMA-BUF CUDA device index fix moved to #2015. This PR is now scoped to
> the topology rewrite + tests, with all reviewer feedback applied.

1. **`[TE]` Replace `getPciDistance` with NCCL-style PCI topology and
   bandwidth-aware NIC selection.**
   The old metric was a string-prefix-of-sysfs heuristic that broke on
   GB200/GB300 (GPUs and NICs on separate PCI domains share only
   `/sys/devices/pci000` as common prefix → flat artifact distances → only
   1 NIC per NUMA selected as preferred). Replaced with NCCL-style PCI
   ancestor-chain walking + path-type classification (PIX / PXB / PHB /
   NODE / SYS) and a 3-level NIC ranking (path type → hops → PCIe bandwidth
   from `current_link_speed * current_link_width`). The bandwidth tiebreaker
   auto-filters lower-bandwidth front-end NICs to `avail_hca` without
   requiring `MC_CUSTOM_TOPO_JSON` or a manual blacklist. Pure-logic helpers
   (`PciPathType`, `InfinibandDevice`, `normalizePciBusId`,
   `classifyGpuNicPath`, `selectPreferredHcas`) moved to
   `include/pci_topology.h` for testability and shared use between
   `discoverCudaTopology` and the unit tests.

2. **`[TE]` Add 23 PCI topology unit tests.**
   Hardware-free tests in `pci_topology_test.cpp` covering
   `normalizePciBusId` (5), `classifyGpuNicPath` (10 — all 5 path types
   plus edge cases), NIC selection ranking via the production
   `selectPreferredHcas` helper (6), and fallback paths (2).

### Reviewer feedback addressed

- Selection ranking extracted into `selectPreferredHcas()` in
  `pci_topology.h`; tests now exercise production code instead of a
  duplicate copy.
- HCA ancestor chains pre-computed once before the GPU loop in
  `discoverCudaTopology` (was O(N_gpu × N_hca) sysfs walks, now
  O(N_gpu + N_hca)).
- Redundant `<ctype.h>` removed from `topology.cpp` (kept `<cctype>`
  since the new code uses `std::tolower` / `std::isxdigit`).

## Module

- [x] Transfer Engine (`mooncake-transfer-engine`)

## Type of Change

- [x] Bug fix
- [x] Refactor

## How Has This Been Tested?

- **Unit tests:** 23 new tests in `pci_topology_test.cpp` (no hardware /
  sysfs dependency). All pass via `ctest -R pci_topology`.
- **GB300, single-host VRAM read (GPU 1 → GPU 0, auto-discovery, no custom
  topo):** 41.22 GB/s @ 64 MB block. The full 18 → 41 GB/s improvement
  requires this PR *together with* the companion DMA-BUF fix (#2015) —
  without that fix, only NICs at verbs index 0–3 pass DMA-BUF validation;
  with both PRs applied, all 10 NICs register and the topology rewrite
  promotes 8 of them to `preferred_hca`.

  | BlkSize | BW (GB/s) | Avg Lat (μs) | P99 Tx (μs) |
  |---|---|---|---|
  | 4 KB    | 0.51  | 8.1    | 8    |
  | 64 KB   | 6.10  | 10.7   | 11   |
  | 1 MB    | 29.10 | 36.0   | 36   |
  | 16 MB   | 40.27 | 416.7  | 421  |
  | 64 MB   | 41.22 | 1628.0 | 1635 |

- **GB300, same-node DRAM, auto-discovery:** 41.34 GB/s @ 64 MB block;
  bandwidth tiebreaker correctly demotes Gen4 front-end NICs to
  `avail_hca`.
- **GB200 (2 GPUs, 6 NICs, same-node DRAM):** auto-discovery with all 6
  NICs reaches 35.5 GB/s; flat topo with 4 backend NICs reaches 39.3 GB/s.

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have formatted my own code using `./scripts/code_format.sh` before submitting.
- [ ] I have updated the documentation. *(no user-facing API change; the
      bandwidth tiebreaker removes the need for `MC_CUSTOM_TOPO_JSON` on
      GB200/GB300, but the env var continues to work for manual overrides)*
- [x] I have added tests to prove my changes are effective.

### CODE REVIEW

File: mooncake-transfer-engine/src/transport/rdma_transport/rdma_context.cpp [hunk 1]
```diff
@@ -582,25 +584,78 @@ int RdmaContext::openRdmaDevice(const std::string &device_name, uint8_t port,
...
 
 #if !defined(WITH_NVIDIA_PEERMEM) && defined(USE_CUDA)
-        // Verify dmabuf support which is required if not using nvidia-peermem.
-        // Assume device index matches.
-        CUdevice cuDevice;
-        CUresult result = cuDeviceGet(&cuDevice, i);
-        if (result != CUDA_SUCCESS) {
-            LOG(ERROR) << "Failed to query CUDA device";
-            return ERR_CONTEXT;
+        // Verify DMA-BUF support against the CUDA device(s) that the local
```
Comment (alogfans) [inline mooncake-transfer-engine/src/transport/rdma_transport/rdma_context.cpp]: I'm not sure whether other (domestic) GPU vendors could support this. Apply this update to NVIDIA only.

**Comment by staryxchen** [comment]
I suggest splitting this into multiple  PRs, for two reasons: 1. It makes review easier; 2. We typically use squash merges, which combine multiple commits into a single commit in main branch, making it difficult to bisect them later.

**Comment by Nishikant-K-P** [comment]
@staryxchen just following up on the review

**Comment by staryxchen** [comment]
> @staryxchen just following up on the review

Sorry for late reply. Could you push an empty commit to re-run CI? And update PR title to reflect actual code changes.
